### DISCUSSION TITLE
Teleport ARMv6 build

### DISCUSSION THREAD

**Discussion by rphanikrishna** [gravitational/teleport#7219] [Q&A] [answered]
Did anyone try or build from source for ARMv6 (Raspberry Pi Zero).

**Comment by webvictim** [comment]
I tried this by setting `GOARCH=arm GOARM=6` in the Teleport build scripts and it seemed to work, so theoretically you could fork the repo and edit the build scripts to add that.

**Reply by despaul** [reply]
Which version of teleport were you able to cross compile to work on raspberry pi with armv6? I keep getting segmentation fault even with the flags above.

**Reply by webvictim** [reply]
I believe it would have been 6.0 or 6.1.

**Reply by despaul** [reply]
Thanks for the quick reply. I just tried  v6.0.0, v6.1.0, v6.1.1, and v6.2.11 without any luck. I am getting either "Segmentation fault" or "Illegal Instruction" on the pi zero with armv6. The host computer used to cross compile is running ubuntu 20.04.1:

GOOS=linux GOARCH=arm CGO_ENABLED=1 CC=arm-linux-gnueabihf-gcc GOARM=6 go build -tags "   webassets_embed" -o build/teleport  -ldflags '-w -s' ./tool/teleport
GOOS=linux GOARCH=arm CGO_ENABLED=1 CC=arm-linux-gnueabihf-gcc GOARM=6 go build -tags "   " -o build/tctl  -ldflags '-w -s' ./tool/tctl
GOOS=linux GOARCH=arm CGO_ENABLED=1 CC=arm-linux-gnueabihf-gcc GOARM=6 go build -tags " " -o build/tsh  -ldflags '-w -s' ./tool/tsh

**Reply by webvictim** [reply]
To be fair, I just tested that it compiled - I didn’t have any hardware to test the resulting build on.

**Reply by despaul** [reply]
Got it! I will post the steps here once I have it running.

**Reply by despaul** [reply]
The only way I was able to build a working version of teleport for armv6 was to compile on a raspberry pi zero w with armv6l.
pi@dppizero:~/teleport $ uname -a
Linux dppizero 5.10.17+ #1414 Fri Apr 30 13:16:27 BST 2021 armv6l GNU/Linux

pi@dppizero:~/teleport $ go env
GO111MODULE=""
GOARCH="arm"
GOBIN=""
GOCACHE="/home/pi/.cache/go-build"
GOENV="/home/pi/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="arm"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/pi/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/pi/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/tmp/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/tmp/go/pkg/tool/linux_arm"
GOVCS=""
GOVERSION="go1.17"
GCCGO="gccgo"
GOARM="6"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/pi/teleport/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -marm -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build1431759178=/tmp/go-build -gno-record-gcc-switches"

Build from master (will try different branches later)
git clone https://github.com/gravitational/teleport.git
cd teleport

pi@dppizero:~/teleport $ make full
---> Copying OSS web assets.
make all WEBASSETS_TAG="webassets_embed"
make[1]: Entering directory '/home/pi/teleport'
---> Building OSS binaries.
make build/teleport build/tctl build/tsh
make[2]: Entering directory '/home/pi/teleport'
GOOS=linux GOARCH=arm CGO_ENABLED=1 CC=arm-linux-gnueabihf-gcc go build -tags "   webassets_embed" -o build/teleport  -ldflags '-w -s' ./tool/teleport
GOOS=linux GOARCH=arm CGO_ENABLED=1 CC=arm-linux-gnueabihf-gcc go build -tags "   " -o build/tctl  -ldflags '-w -s' ./tool/tctl
GOOS=linux GOARCH=arm CGO_ENABLED=1 CC=arm-linux-gnueabihf-gcc go build -tags " " -o build/tsh  -ldflags '-w -s' ./tool/tsh
make[2]: Leaving directory '/home/pi/teleport'
make[1]: Leaving directory '/home/pi/teleport'
pi@dppizero:~/teleport $ cd build
pi@dppizero:~/teleport/build $ ls
tctl  teleport  tsh
pi@dppizero:~/teleport/build $ file teleport 
teleport: ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-armhf.so.3, for GNU/Linux 3.2.0, Go BuildID=YWK0gu_RmVH90LYJv01o/-itSFQkc8unmcmbOLRAE/2K-kGg5ICj5-oPn6Aw8W/ysG60ka86GKsuFpC4NaK, BuildID[sha1]=28bc653e354eb68845985e437e4cc88431208ab2, stripped
pi@dppizero:~/teleport/build $ ./teleport 
Usage: teleport [<flags>] <command> [<args> ...]

Clustered SSH service. Learn more at https://goteleport.com/teleport

Flags:

Commands:
  help         Show help.
  start        Starts the Teleport service.
  status       Print the status of the current SSH session.
  configure    Generate a simple config file to get started.
  version      Print the version.
  app start    Start application proxy service.
  db start     Start database proxy service.

**Reply by webvictim** [reply]
Interesting. I suspect there likely is a way to cross-compile this for ARMv6 on a different machine, but it might need a different EABI layer installed.

I'm wondering what the difference looks like between `file ./teleport` for the binary compiled direct on the Pi zero vs the cross-compiled binary?

**Reply by hugoShaka** [reply]
This is a GCC limitation: https://stackoverflow.com/questions/55465118/gcc-8-cross-compiler-outputs-armv7-executable-instead-of-armv6
The C compiler is outputting armv7 bytecode instead of armv6. Some people seem to have worked around by building custom glibc and GCC:
- https://gist.github.com/tttapa/534fb671c5f6cced0e1722d3e4aec987
- https://github.com/Pro/raspi-toolchain/blob/master/Dockerfile

**Comment by garsiv1932** [comment, marked as answer]
Hello people, i made the cross-compiling for an ARM V6 (Raspi Zero). What I did:

* Create a VM on VirutalBox as the compilation process use the libc library, and if you compile in ubuntu, then you need to have the same version of libc library in the raspi zero host, and because I use raspbian, the libc library of Debian is older than the ubuntu one, so, if you are a Ubuntu user, you need to do the process in a Debian VM
* Second step is to download this cross compiler:
 https://master.dl.sourceforge.net/project/raspberry-pi-cross-compilers/Raspberry%20Pi%20GCC%20Cross-Compiler%20Toolchains/Buster/GCC%208.3.0/Raspberry%20Pi%201%2C%20Zero/cross-gcc-8.3.0-pi_0-1.tar.gz
Unzip it in /opt.
* Change the path directory in the VM default user profiles file ($HOME/.profile) and put this line at the end 
PATH=$PATH:/opt/cross-pi-gcc-8.3.0-0/bin
* Compile the app with this command
GOS=linux GOARCH=arm GOARM=6 make full

And wuala, you can move your binaries to your raspi zero and they will work.

People of teleport, you can put this compiler on the pipeline and create .deb file for armv6. Raspi Zero is popular on IOT as it's very cheap, and I use teleport to access my remote raspis which connects for 3 and 4g network connections.

I expect this can be useful for others.
@hugoShaka sorry for the delay in this answer, I had some busy weeks.

**Reply by webvictim** [reply]
Thanks for the write-up!
